### ISSUE TITLE
Improved A11y Support - TabIndex

### ISSUE THREAD

**Issue by samhouts** [xamarin/Xamarin.Forms#2789]
## Description
As an app developer, I currently am unable to specify an expected focus/tab order for elements in my Xamarin.Forms app that will be used by a screen reader and/or keyboard navigation to improve the usability and accessibility of the app.  When this spec is complete, I can explicitly set the `TabIndex` on any focusable element to have it read to the app user and/or prepare it for input in a particular order.

### Acceptance criteria
* Solution must enable app developer to set an integer value on `Element.TabIndex` with the expectation that focus will cycle through each element in ascending order. 
* Conflicts on `TabIndex` will be resolved by the declaration order. 
* If no value is set in `TabIndex`, the default behavior will be used. 
* If the app developer attempts to set a focus order using a custom renderer, the values set by the custom renderer should take precedence.

#### Per the UWP specification:
* UI elements with `TabIndex` equal to 0 are added to the tab order based on declaration order in XAML or child collections.
* UI elements with `TabIndex` greater than 0 are added to the tab order based on the `TabIndex` value.
* UI elements with `TabIndex` less than 0 are added to the tab order and appear before any zero value.

## API changes
Add: `int Element.TabIndex { get; set; }`

### Description of changes
- UWP: https://docs.microsoft.com/en-us/windows/uwp/accessibility/keyboard-accessibility
- iOS: https://developer.apple.com/documentation/objectivec/nsobject/1615147-accessibilityelements#declarations 
- Android: https://developer.android.com/training/keyboard-input/navigation.html#Tab 
- macOS: https://github.com/xamarin/Xamarin.Forms/pull/1060 (convert from platform specific)

**Comment by StephaneDelcroix** [comment]
for elements with equal TabIndex, declaration order matters, right ?

**Comment by samhouts** [comment]
@StephaneDelcroix  Yep:

> Conflicts on TabIndex will be resolved by the declaration order.

**Comment by GalaxiaGuy** [comment]
Will this have any impact on directional navigation (using a hardware dpad)?

The Android link suggests basic dpad navigation should work by default, but it does not seem to in a Xamarin Forms app:
https://bugzilla.xamarin.com/show_bug.cgi?id=30891

(I haven't yet tried other platforms.)

**Comment by johnshardman** [comment]
Currently (ok, in XF 2.4.x), tabbing through the controls on UWP results in a sequence where the tab position is unknown for three tabs (i.e. after tabbing off the last control on a page, and before cycling back to the first control on a page, three extra tab presses have to be done). Will this change get rid of those three unknown tab stops? Also, will this change put the navigation bar's back button on the tab sequence? Without that (unless there's some key combination I don't know about), it's possible to navigate on a page using the keyboard, but not go back to a previous page using the back button.
